[Frontend] Align tool_choice="required" behavior with OpenAI when tools is empty

n0gu-furiosa opened the pull request:
## Purpose

When a user specifies `tool_choice="required"` with an empty `tools`, the OpenAI-compatible server does not currently handle this scenario in the most appropriate way. The response returned is as follows:

```text
$ curl http://localhost:8000/v1/chat/completions -H 'Content-type: application/json' -d '
{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [{"role": "user", "content": ""}],
    "tool_choice": "required",
    "tools": []
}'
{"object":"error","message":"Grammar error: Unsatisfiable schema: required item is unsatisfiable","type":"BadRequestError","param":null,"code":400}
```

This `BadRequestError` is not raised by the model validator but by the guided decoding backend, because the request generates an unsatisfiable JSON schema:

```
{"type":"array","minItems":1,"items":{"type":"object","anyOf":[]}
```

Although this behavior technically blocks invalid requests, it is not the most user-friendly. The issue becomes more problematic when `stream=True`: in this case, the request is not validated before upgrading to SSE. Instead, the error surfaces after the model has begun generating results:

```
$ curl http://localhost:8000/v1/chat/completions -H 'Content-type: application/json' -d '
{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [{"role": "user", "content": ""}],
    "tool_choice": "required",
    "tools": [],
    "stream": true
}'
data: {"error": {"object": "error", "message": "Grammar error: Unsatisfiable schema: required item is unsatisfiable", "type": "BadRequestError", "param": null, "code": 400}}

data: [DONE]
```

This PR adds explicit validation of tool length when `tool_choice="required"`.

## Test Plan

Manually ran `vllm serve meta-llama/Llama-3.1-8B-Instruct --enable-auto-tool-choice --tool-call-parser llama3_json` and executed the curl command above.

## Test Result

```
$ curl http://localhost:8000/v1/chat/completions -H 'Content-type: application/json' -d '
{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [{"role": "user", "content": ""}],
    "tool_choice": "required",
    "tools": [],
    "stream": true
}'
{"object":"error","message":"[{'type': 'value_error', 'loc': ('body',), 'msg': 'Value error, When using `tool_choice` as \"required\", `tools` must contain at least one tool', 'input': {'model': 'meta-llama/Llama-3.1-8B-Instruct', 'messages': [{'role': 'user', 'content': ''}], 'tool_choice': 'required', 'tools': [], 'stream': True}, 'ctx': {'error': ValueError('When using `tool_choice` as \"required\", `tools` must contain at least one tool')}}]","type":"Bad Request","param":null,"code":400}
```

n0gu-furiosa committed:
[Frontend] Validate tool list length when tool_choice=required

Signed-off-by: Sungyoon Jeong <sungyoon.jeong@furiosa.ai>

chaunceyjiang reviewed (commented):
Could you help verify how the OpenAI online service handles this situation?

n0gu-furiosa commented:
@chaunceyjiang I checked, and it turns out that the OpenAI API behaves as if `tool_choice` were `"none"` when tools is an empty list (or `"auto"`, I believe they behave the same way in this context).

Request payload:
```json
{
    "model": "gpt-4o",
    "messages": [
        {
            "role": "user",
            "content": "How is the weather today in Seoul?"
        }
    ],
    "stream": false,
    "tool_choice": "required",
    "tools": []
}
```

Response body:
```json
    "choices": [
        {
            "index": 0,
            "message": {
                "role": "assistant",
                "content": "I currently don't have real-time weather data. You might want to check a reliable weather website or use a weather app for the most current information in Seoul.",
                "refusal": null,
                "annotations": []
            },
            "logprobs": null,
            "finish_reason": "stop"
        }
    ],
```

To align with this behavior, I think the simplest approach would be to override `tool_choice` to `"none"` if `tool_choice` is `"required"` and `tools` is an empty list, since much of the code assumes that tool_choice `"required"` must return tool calls. If you agree, I can update the PR this way. Do you have any thoughts or concerns?

chaunceyjiang commented:
Sounds good.
/cc @aarnphm

n0gu-furiosa committed:
Treat tool_choice="required" and empty tools as equivalent to "none"

Signed-off-by: Sungyoon Jeong <sungyoon.jeong@furiosa.ai>

n0gu-furiosa commented:
@chaunceyjiang Applied in [781573d](https://github.com/vllm-project/vllm/pull/21052/commits/781573d747aa2304c11392ade70534e0e2ac4cc1). Also changed the title of this PR.

Updated test result:

```
$ curl http://localhost:8000/v1/chat/completions -H 'Content-type: application/json' -d '
{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [{"role": "user", "content": "How is the weather today in Seoul?"}],
    "max_completion_tokens": 20,
    "tool_choice": "required",
    "tools": []
}'
```

```
{
  "id": "chatcmpl-40d600203d804754a8d4c58d4cae5c8b",
  "object": "chat.completion",
  "created": 1752726568,
  "model": "meta-llama/Llama-3.1-8B-Instruct",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "I'm not able to access real-time information. However, I can suggest some ways to find the",
        "refusal": null,
        "annotations": null,
        "audio": null,
        "function_call": null,
        "tool_calls": [],
        "reasoning_content": null
      },
...
```

n0gu-furiosa commented:
Hi @chaunceyjiang, just a friendly reminder to take a look at this when you have a chance. Thanks in advance.

chaunceyjiang reviewed (approved):
Thanks~

LGTM.

/cc @DarkLight1337 @aarnphm

n0gu-furiosa committed:
Merge branch 'vllm-project:main' into tools-required-validation

n0gu-furiosa committed:
Merge branch 'vllm-project:main' into tools-required-validation
